Simplify the code by using GitHub CLI and the GitHub REST API

seanh opened the pull request:
...also add tests. 

This results in significantly less than half of the original lines of code (if you don't count the new tests) by:

* Removing the Python Slack API client in favour of the `slackapi/slack-github-action`.  
  This change has already been done on `main`.

* Removing the GitHub API client in favour of shelling out to GitHub CLI.

* Calling GitHub's REST API instead of their GraphQL API.

* The output format has also been simplified.

I think we should probably prefer the REST API over GraphQL whenever possible: it's both simpler and better documented. If this breaks because GitHub change their API, I'd much rather have to update a script's REST API usage than try to figure out what broke with a GraphQL query.

The tests may seem a bit OTT but they will be helpful if we want to make changes to this in future. If they become a hindrance we can always just delete them.

## Testing manually

```terminal
$ tox -qqe dev --run-command 'dependabot-alerts hypothesis'
```

This will use your GitHub CLI authentication (set up via `gh auth login`) or you can set `GITHUB_TOKEN` to use a different authentication token.

The output format is simple: one alert per line, with no preamble or footer lines or indentation etc. This is good UNIX command design, for example it makes the output easy to parse in case you want to pipe it to another command, for example to count the number of alerts: `dependabot-alerts hypothesis | wc -l`.

There's also:

```terminal
$ dependabot-alerts --help
$ dependabot-alerts --version
```

## Deduplication

We get the same alert multiple times in the same repo when the same package is mentioned in multiple Python requirements files. When this happens only one line will be output but it'll note the number of duplicate copies:

    hypothesis/somerepo GHSA-****-****-**** alerts:7 severity:medium <package_name> "<security_advistory_summary>" https://github.com/hypothesis/somerepo/security/dependabot/123

Dependabot security alerts have a GHSA ID field that is a required, non-nullable unique ID. There are other IDs in there as well, like the CVE ID, but these are optional or nullable. The code considers two alerts to be equal if they belong to the same repo and have the same GHSA ID, and will group them onto one line as above.

## Slack formatting

Pass `--slack` to print output in the form of a nicely-formatted Slack post:

```terminal
$ tox -qqe dev --run-command 'dependabot-alerts --slack hypothesis'
```

## When there are no alerts

If there are no alerts then `dependabot-alerts hypothesis` will just exit with `0` and print nothing. Again: good UNIX command design. For example `ls` doesn't print `There are no files` if you run it in an empty directory.

With `--slack` if there are no alerts it'll print `There are no Dependabot security alerts in the hypothesis GitHub organization` and the GitHub Actions workflow will actually post that message into Slack. This might be useful, as a reminder that the scheduled action is running? Alternatively we could make it print nothing in this case and then make the workflow not try to post to Slack if there is no output.

## Error handling

If something goes wrong you'll get an error message from GitHub CLI and it'll exit with `1`. It's not super friendly but it's fine. For example if you're not logged in:

```terminal
$ tox -qqe dev --run-command 'dependabot-alerts hypothesis'

To get started with GitHub CLI, please run:  gh auth login
Alternatively, populate the GH_TOKEN environment variable with a GitHub API authentication token.

Traceback (most recent call last):
...
```

robertknight commented:
> The output format is simple: one repo/package per line, with no preamble or footer lines or indentation etc. This is good UNIX command design, for example it makes the output easy to parse in case you want to pipe it to another command, for example to count the number of vulnerable packages: dependabot-alerts hypothesis | wc -l.

The output in the earlier version was designed to be useful for a human to read and act on. I don't mind having a succinct pipeline-friendly mode as an option, like many CLI utilities do (or check `isatty`), but it is annoying not to have this information in the output:

- Convenient link to alert
- Severity
- Short summary of what the issue is

> The code on main seems to have its own concept of "vulnerabilities" that doesn't seem to map directly to what the Dependabot alerts REST API returns, and it seems to only report two vulnerabilities in Via HTML. I'm not sure I understand what the code on main is doing there.

Each `Vulnerability` object on main corresponds to one alert from Dependabot. The name comes from the [`RepositoryVulnerabilityAlert`](https://docs.github.com/en/graphql/reference/objects#repositoryvulnerabilityalert) schema in GraphQL), but with a limit that only one is reported per (package, repository).

The rationale is that in a given repository, you would normally only need to take one action per affected package, rather than for each alert if there are multiple vulnerabilities in the same package or multiple alerts for the same issue that affect different lockfiles.

> `Message generated by the `alerts.yml` workflow <https://github.com/hypothesis/dependabot-alerts/blob/main/.github/workflows/alert.yml|in dependabot-alerts>`

This is useful, although it would be nice if this information could be added automatically by the thing that posts to Slack.

> If there are no alerts then dependabot-alerts hypothesis will just exit with 0 and print nothing. Again: good UNIX command design. For example ls doesn't print There are no files if you run it in an empty directory.

This is fine, though I do use plenty of tools which have decided to show something when nothing needs doing and instead supply a silent/quiet flag to suppress these (cf black, MyPy). The main need for having a message when there are no alerts is in the case when reporting to Slack, to serve as a confirmation that the check is happening.

seanh commented on src/dependabot_alerts/core.py:
```diff
@@ -1,184 +1,63 @@
+
+        for alert_dict in alert_dicts:
+            alert = Alert.make(alert_dict)
+            if alert in alerts:
+                alerts[alert].duplicates.append(alert)
+            else:
+                alerts[alert] = alert
```
This is the deduplication here

seanh commented:
> The output in the earlier version was designed to be useful for a human to read and act on. I don't mind having a succinct pipeline-friendly mode as an option, like many CLI utilities do (or check `isatty`), but it is annoying not to have this information in the output:
> 
>     * Convenient link to alert
> 
>     * Severity
> 
>     * Short summary of what the issue is

I've added each alert's URL, severity and summary to the output (both Slack and non-Slack).

The Slack ones should be quite readable and easy to act on by clicking a link.

We could also add a `--pretty` for the local terminal version if you want.

> Each `Vulnerability` object on main corresponds to one alert from Dependabot. The name comes from the [`RepositoryVulnerabilityAlert`](https://docs.github.com/en/graphql/reference/objects#repositoryvulnerabilityalert) schema in GraphQL), but with a limit that only one is reported per (package, repository).
> 
> The rationale is that in a given repository, you would normally only need to take one action per affected package, rather than for each alert if there are multiple vulnerabilities in the same package or multiple alerts for the same issue that affect different lockfiles.

I've taken a slightly different approach to deduplication on this branch: it groups together alerts that belong to the same repo and have the same GHSA ID. So you might still get multiple alerts for the same repo+package if there are actually two different alerts for the same package, but not if it's just the same alert for different lockfiles. We can easily change this deduplication choice to be based on repo+package instead.


> `Message generated by the `alerts.yml` workflow <https://github.com/hypothesis/dependabot-alerts/blob/main/.github/workflows/alert.yml|in dependabot-alerts>`
> 
> This is useful, although it would be nice if this information could be added automatically by the thing that posts to Slack.

I think it'd be nice to add a reusable `slack.yml` workflow to https://github.com/hypothesis/workflows. That would have other benefits as well (easier to have Dependabot keep the version number of the Slack action up to date) and the reusable workflow could add this "generated by..." (referencing the caller workflow), then all our Slack alerts would get it. For another day perhaps.
 
> > If there are no alerts then dependabot-alerts hypothesis will just exit with 0 and print nothing. Again: good UNIX command design. For example ls doesn't print There are no files if you run it in an empty directory.
> 
> This is fine, though I do use plenty of tools which have decided to show something when nothing needs doing and instead supply a silent/quiet flag to suppress these (cf black, MyPy). The main need for having a message when there are no alerts is in the case when reporting to Slack, to serve as a confirmation that the check is happening.

I've added this "There are no alerts" confirmation message when `--slack` is used.

robertknight reviewed (approved):
Thanks for this overhaul and the tests. Let's get this landed and we can discuss what to do about the current open alerts in ViaHTML.

seanh commented on src/dependabot_alerts/cli.py:
```diff
@@ -1,49 +1,43 @@
+        for alert in alerts:
+            print(
+                f'{alert.repo_full_name} {alert.ghsa_id} alerts:{len(alert.duplicates) + 1} severity:{alert.severity} {alert.package} "{alert.summary}" {alert.html_url}'
+            )
```
I've pushed a refactoring that separates each formatting option into a separate `format_*()` function in `format.py` and changes the `--slack` option to `--format slack`. This should make it easy to add more formats like `--format pretty` or `--format json` etc without the code quality deteriorating.

* https://github.com/hypothesis/dependabot-alerts/blob/468bea0d554cb3a9fdbe8c71a6395bd48ef2edaf/src/dependabot_alerts/format.py
* https://github.com/hypothesis/dependabot-alerts/blob/468bea0d554cb3a9fdbe8c71a6395bd48ef2edaf/src/dependabot_alerts/cli.py#L22-L29

seanh commented on src/dependabot_alerts/core.py:
```diff
@@ -1,184 +1,66 @@
+            )
+        except CalledProcessError as err:  # pragma: no cover
+            print(err.stdout)
+            print(err.stderr)
```
I've changed this to print any stdout from GitHub CLI to stdout and any stderr from GitHub CLI to stderr, and to prefix each line with `GitHub CLI stdout> ` or `GitHub CLI stderr> `:

https://github.com/hypothesis/dependabot-alerts/blob/8046836b6eda14dcd8091037f80708c2079c9689/src/dependabot_alerts/core.py#L54-L59

Example (trying to get alerts for a user rather than an organization fails):

```terminal
$  tox -qqe dev --run-command 'dependabot-alerts seanh'
GitHub CLI stdout> {"message":"Not Found","documentation_url":"https://docs.github.com/rest/dependabot/alerts#list-dependabot-alerts-for-an-organization"}
GitHub CLI stderr> gh: Not Found (HTTP 404)
Traceback (most recent call last):
  File "/home/seanh/GitHub/hypothesis/dependabot-alerts/.tox/dev/bin/dependabot-alerts", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/seanh/GitHub/hypothesis/dependabot-alerts/.tox/dev/lib/python3.12/site-packages/dependabot_alerts/cli.py", line 22, in cli
    repos = github.alerts(organization)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/seanh/GitHub/hypothesis/dependabot-alerts/.tox/dev/lib/python3.12/site-packages/dependabot_alerts/core.py", line 43, in alerts
    result = self._run(
             ^^^^^^^^^^
  File "/home/seanh/.pyenv/versions/3.12.0/lib/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['gh', 'api', '--paginate', '/orgs/seanh/dependabot/alerts?state=open']' returned non-zero exit status 1.
```
